### ISSUE TITLE
Docs improvements: Decoupling from framework releases 

### ISSUE THREAD

**Issue by stichbury** [kedro-org/kedro#2349]
## Introduction
Something that @astrojuanlu and I discussed recently: docs are tied to releases so all the useful little tweaks and updates that we put in to fix things are suspended until occasions when a release happens and stable is incremented.

## Background to the problem
We've seen this particularly with Kedro-Viz in the last few days: the docs have been ready to go, and the release announcement points to them, but because the release isn't out, things are blocked. 

I've seen this in other projects too and, from an authoring point of view, it's just frustrating not to be able to push content improvements out because they are coupled to other parts of the system, even when they aren't related to them.

We think it may make sense to make the default version of the docs, the latest rather than the stable version. That is, the latest docs on main are the ones that are presented when someone navigates to [kedro.readthedocs.io](http://kedro.readthedocs.io/).

### Advantages
Anything accepted onto main through a PR gets published immediately as a default version and we uncouple the docs from the release process.

### Disadvantages
Any docs about unreleased changes need to be "supported" until a release is made (that is, we may have queries about things that aren't yet available, if we document them)

I think this is a reasonable summary...maybe we could give it a trial run between 0.18.5 and 0.19? (edited)

**Comment by stichbury** [comment]
Comments on this (we discussed on Slack originally)

@AntonyMilneQB 

I had exactly the same idea some time last year. I spoke to Lorena about it and she said that actually this is the way that the docs used to be originally , and people got confused for exactly the reason you suggest: they were sometimes looking at docs for unreleased features.
Personally I would support the change since I think the advantage outweighs the disadvantage. At least I think we should give it a trial and see if it causes any problems. (edited) 


@stichbury 
 
Thanks! I can see an issue with docs for unreleased features but we should be able to "write around" most cases - and if people start raising questions because something documented isn't yet available, we get to see who is interested in the feature). I think the advantages of unblocking the docs publication are much more significant.

@noklam 
I would support it if there are ways to separate the docs between the API doc and the rest of it.
I checked a few major data science libraries (i.e. keras, pytorch , scikit-learn, tensorflow), and they all pin the stable version by default.

I think this aligns with my experience as a user, I expect the docs to be something that is ready to use, If I am looking for unstable changes, I will look for it specifically so I am very aware of what I am looking for.
With that said I do understand sometimes the docs change is not seen for weeks or even months. Should we have more frequent minor releases instead? (edited) 

@astrojuanlu 

Looking at https://github.com/kedro-org/kedro/tags, our release cadence is ~2-3 months. since we're at 0.x still, the concept of minor vs micro releases gets a bit blurry... but I don't think we strictly follow SemVer anyway ("major version zero [means] Anything MAY change at any time").
scikit-learn releases micro versions on demand, for example 1.1.0 was released in May 12 and 1.1.1 in May 19. 

Asking @noklam It looks like that have never happened for Kedro - is it because our release policy discourages that, or because we've never had a bug we wanted to patch quickly?
if we could have more frequent* micro releases†, and treat certain docs improvements as bugs we want to patch (for our readers to benefit from better docs) it would alleviate our pains somewhat while also keeping the docs pinned to stable .
*on request? every month? not sure how much overhead does a release have - but if releasing is painful, maybe that's something we should address separately)
†0.18.5.1? having 4 segments already feels like too much, but maybe we want to keep our "micro" meaning "minor"

@noklam 

I think we do have occasions when we want to patch things quicker. i.e. the release 0.18.0 is on 1 Apr and 0.18.1 is on 10 May.
I think there are a few reasons for this but @merelcht can supplement more

Users usually don’t upgrade immediately after a new release comes out ( In a sense unless it’s a quick bug fix, these micro release almost sounds like doc release to me)
We usually work in sprints so it’s a 2 weeks chunk of time, any bug we found within this sprint, will need to wait until the next Backlog Grooming, go through discussion, and then schedule in the next release. So this takes ~3 weeks or if it’s one sprint later then it is already 5 weeks. I think it’s good to have this process, although it means our response time is weeks, not days.
There are indeed some release overheads as we need to coordinate what gets into the release. If we wanna do more frequent releases, it also means that we need to accept some incomplete features that will get into the releases.


@noklam 
That said I don’t mind a more frequent release schedule, for example, the release for 0.18.3, the urgent [bug fix](https://github.com/kedro-org/kedro/commit/0a44c5aa57b022e1f296b7f855c14634fa88b742) (In this case a CST was waiting for our fix) was done on 27 Aug, but the release is out on 20 Sep as we are waiting for other features to be completed.

@astrojuanlu 
thanks for the extra context. The problem I see with making quicker 0.18.X releases is that it impacts the workflow of everybody else, while fixing a typo in the docs or restructuring a guide sounds like something that could be done more or less transparently with a different process. I see several things we can do:

* docs point to latest by default. cons already described (for periods of time, users might find docs for unreleased features without noticing). could be amended by adding an admonition on top or even more prominent visual elements, but that doesn't reduce the confusion risk to zero.
* docs point to stable but releases are quicker. impacts all the team, there should be consensus on this.
we separate API docs from narrative docs. pros: achieves decoupling, and gives us freedom to break from Sphinx. cons: infra overhead, and requires designing the docs in a way that going back and forth between the narrative and the API ones is easy.

@AntonyMilneQB 
This is a tricky one because I agree with everything that’s been said above and don’t have any better ideas…
Even though I think we should aim to do faster patch releases (agree that 4 segments feels like too much, so these would remain 0.18.x), having the docs update process ultimately limited by this still doesn’t feel great.
Seeing as docs are no longer packaged with kedro (which they used to be), I don’t see any reason we shouldn’t in theory do option 3 of the above and publish non-API docs independently of a kedro release. I have no idea how this would work in practice though, do you @astrojuanlu? Since all docs (narrative + API) are versioned in the same way on RTD I guess we’d have to have something that only rebuilds the API docs on release, so that latest points to latest narrative + stable API. But then maybe that itself is confusing because it’s not consistent :confused: (edited) 

@yetudada 
I'll leave a few comments to consider here and thank you for trying to solve this problem.

TL;DR We should release more often, the cadence should be after every sprint - addressing the root cause of the problem.
Here's additional comments as to why I recommend this:

* Idea 1: I was around when docs pointed to latest and we did get queries about unreleased features. We should always present a stable, and "working version" of Kedro to limit unexpected errors for users.
* Idea 2: This addresses the root cause of the problem. I'm happy to release after every sprint.
* Idea 3: I'm not sure if this solves the problem of making sure that the documentation is up to date because the docs are still coupled to the release - especially if new functionality is written in the narrative docs. However, we do have an outstanding task of doing this work because of our issues with Sphinx and the complications around our current documentation build because of kedro.datasets.

@merelcht 
I 100% agree we should release more frequently. I think because we’re used to pretty large minor releases every couple of months, we make it a very big thing in our heads and also try to add on a 1000 things before we release. It’s okay for releases to only have a couple of changes.

**Comment by astrojuanlu** [comment]
Notice that we might still explore workflow changes to address some tooling pains with Sphinx as well as attempt to consolidate the documentation of our several packages (see gh-2072), but closing this issue means that release cadence won't be a factor to consider.
